rumqttc/src/lib.rs

```diff
@@ -624,11 +613,44 @@ impl MqttOptions {
     /// When set `false`, broker will hold the client state and performs pending
     /// operations on the client when reconnection with same `client_id`
     /// happens. Local queue state is also held to retransmit packets after reconnection.
+    ///
+    /// NOTE: It panicks if the `client_id` is a zero byte string and `clean_session` is
+    /// not `true`.
+    ///
+    /// ```should_panic
+    /// # use rumqttc::MqttOptions;
+    /// let mut options = MqttOptions::new("", "localhost", 1883);
+    /// options.set_clean_session(false);
+    /// ```
     pub fn set_clean_session(&mut self, clean_session: bool) -> &mut Self {
+        if self.client_id.is_empty() {
+            assert!(
+                clean_session,
+                "Cannot unset clean session when client id is empty"
+            );
+        }
```

Comment:
assert message can be improved like: "Clean session must be true when client id is provided"

btw, we can use single assert statement instead of two checks:
```rust
assert!(!client_id.is_empty() || clean_session, "Clean session must be true when client id is empty")
```